### PR TITLE
Feat 1385 & 1355

### PR DESCRIPTION
**Pull request by phpgeekfr**
This PR builds on top of feat-1061-bis
Feat 1385 & 1355 : some debug on catalog dashboard + contract implementation

### CODE REVIEW

File: apps/catalog/dashboard/dashboard/src/app/dashboard.module.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { MovieCreateComponent } from '@blockframes/movie/movie/components/movie-
 const routes: Routes = [
   {
     path: '',
-    redirectTo: 'home',
+    redirectTo: 'catalog',
```
Comment (GrandSchtroumpf) [inline apps/catalog/dashboard/dashboard/src/app/dashboard.module.ts]: Why did you remove the home page ?

File: apps/catalog/dashboard/dashboard/src/app/pages/dashboard-home/dashboard-home.component.html [hunk 2]
```diff
@@ -1,8 +1,53 @@
 <section test-id="content" page-id="catalog-homepage">
-    <article fxLayout="column" fxLayoutAlign="center center">
-        <h1 class="mat-display-3">Welcome to Bigger Boat - Dashboard</h1>
-        <p class="mat-body-1">
-            A consolidation of film libraries to create new opportunities through secure transactions.</p>
-        <button mat-button color="accent" test-id="import" [routerLink]="['/layout/o/home/import']">Upload avails & films doc</button>
-    </article>
-</section>
+  <article fxLayout="column" fxLayoutAlign="center center">
```
Comment (GrandSchtroumpf) [inline apps/catalog/dashboard/dashboard/src/app/pages/dashboard-home/dashboard-home.component.html]: Is it supposed to match the new design ? Should it go into the mono-catalog app ?

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+type Timestamp = firestore.Timestamp;
+
+export const enum ContractStatus {
+  submitted = 'submitted',
+  accepted = 'accepted',
+}
+
+export interface ContractTitleDetail {
+  titleId: string, // ie: movieId
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: Why not naming it movieId ?

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+interface ContractRaw<D> {
+  id: string,
+  parentContractIds?: string[],
+  childContractIds?: string[],
+  parties: Party[],
+  status: ContractStatus,
+  scope: TermsRaw<D>,
+  signDate?: D,
+  titles: { [titleId: string]: ContractTitleDetail },
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: Use native typescript typing `Record<string, ContractTitleDetail>`

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  price: Price;
+  paymentSchedule?: string; // @todo #1397 change this when creating invoices
+}
+
+export interface InvoiceRaw<D> {
+  // @todo #1397
+}
+
+export interface ContractDocumentWithDates extends ContractRaw<Date> {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: I really don't like this `WithDates` pattern. Why not just `Contract` for Frontend & `ContractDocument` for Backend. We end up with super long name, and if we have more stuff to change between both, the naming `withDates` won't have sense anymore.

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+export interface CatalogCart {
+  name: string;
+  status: CartStatus;
+  deals: string[];
+  price: Price;
+}
+
+export interface MovieData { // @TODO (#1388) check this object
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Could we not use a subset of the movie object instead ?

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  movieName: string;
+  duration: DateRange;
+  territory: string;
+  rights: string;
+  languages: string;
+  dubbed: string;
+  subtitle: string;
+}
+
+export type Contract = ContractDocumentWithDates;
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: This doesn't change my comment above. I think it's complicated process `ContractRaw -> ContractRaw<Date> -> ContractDocumentWithDates -> Contract`.

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  return {
+    titleId: '',
+    distributionDealIds: [],
+    ...params,
+    price: createPrice(params.price),
+  };
+}
+
+/**
+ * @param contract
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Add comment

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+ */
+export function validateContract(contract: Contract): boolean {
+
+  // First, contract must have at least a licensee and a licensor
+
+  if (contract.parties.length < 2) { return false; }
+  const licensees = contract.parties.filter(p => p.role == getCodeIfExists('LEGAL_ROLES', 'licensee'))
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Is ok if you don't have a lot of contract parties (<1000). Else, better use a single loop

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  const licensors = contract.parties.filter(p => p.role == getCodeIfExists('LEGAL_ROLES', 'licensor'))
+
+  if (!licensees.length || !licensors.length) { return false; }
+
+  for(const licensee of licensees) {
+    if(licensee.orgId === undefined) {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: What about null or empty string ?

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+      delete licensee.orgId
+    }
+    if(typeof licensee.showName !== 'boolean') {
+      return false;
+    }
+  }
+
+  for(const licensor of licensors) {
+    if(licensor.orgId === undefined) {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Same here

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+  }
+
+  //////////////////
+  /// CART STUFF
+  //////////////////
+
+  /**
+   * Adds a deal to cart identified by "name"
+   * @param dealId 
+   * @param name
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: What name is it?

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+   * @param name 
+   */
+  public async addDealToCart(dealId: string, name: string): Promise<CatalogCart> {
+    const cart = await this.getCart(name);
+    cart.deals.push(dealId);
+    return this.updateCart(cart);
+  }
+
+  /**
+   *
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: Comments

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+
+  /**
+   * Adds a deal to cart identified by "name"
+   * @param dealId 
+   * @param name 
+   */
+  public async addDealToCart(dealId: string, name: string): Promise<CatalogCart> {
+    const cart = await this.getCart(name);
+    cart.deals.push(dealId);
+    return this.updateCart(cart);
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: You should use a firestore transaction. 
Use native `akita-ng-fire` instead, it does all the good stuff for you :
```typescript
this.update(name, (cart) => ({ deals: [ ...cart.deals, dealId ]})) 
```

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+  }
+
+  /**
+   * 
+   * @param amount 
+   * @param currency
+   * @param _name
+   */
+  public async submitCart(amount: number, currency: MovieCurrenciesSlug, _name?: string) {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: Same here.

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+      status: CartStatus.submitted
+    };
+    this.updateCart(updatedCart);
+  }
+
+  /**
+   * Creates an empty cart
+   * @param _name
+   */
+  private async initCart(_name?: string): Promise<CatalogCart> {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: Same here.
Also do `initCart(name: string = 'default')`

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+    const name = _name ? _name : 'default';
+    const cart: CatalogCart = createCart({ name });
+    await this.db.doc<CatalogCart>(`orgs/${this.organizationQuery.getActiveId()}/cart/${name}`).set(cart);
+    return cart;
+  }
+
+  private async updateCart(cart: CatalogCart): Promise<CatalogCart> {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: Remove this function if doesn't do anything more than native `akita-ng-fire`

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts (hunk 14 above)
Comment (GrandSchtroumpf) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: You'll need to ceate a getter for your path also (see with Hugo)

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts (hunk 3 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: New denomination. To handle movies, series etc .. Check with Vincent ;)

File: apps/catalog/dashboard/dashboard/src/app/dashboard.module.ts (hunk 1 above)
Comment (phpgeekfr) [reply apps/catalog/dashboard/dashboard/src/app/dashboard.module.ts]: The old one was broken. Just added a new component to be able to use and test the app. This is a quickfix, no css, no specific tests, just a movie list.

File: apps/catalog/dashboard/dashboard/src/app/pages/dashboard-home/dashboard-home.component.html (hunk 2 above)
Comment (phpgeekfr) [reply apps/catalog/dashboard/dashboard/src/app/pages/dashboard-home/dashboard-home.component.html]: Nop, just for testing and to have a "running" app

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts (hunk 5 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: Just did what was already used everywhere. Dont really like this WithDates either :)

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts (hunk 6 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Yes of course it should. Don't know where this object come from, but I think this is going to be deleted in #1388

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts (hunk 9 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: It will never exceed 4 or 5 parties

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts (hunk 12 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: This is the name of the cart. `default` if not specified on creation. This will allow us to handle multiple carts for a unique org. Added comment

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts (hunk 14 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.service.ts]: added comment

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts (hunk 3 above)
Comment (GrandSchtroumpf) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: Maybe you can add a comment `/** movie/serie id */`

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts (hunk 5 above)
Comment (GrandSchtroumpf) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: Could you check with @ioaNikas or @Hugo-Hontans to see if we can improve that. I'm afraid it's become a "thing" ^^

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts (hunk 6 above)
Comment (GrandSchtroumpf) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Can you change the `TODO` naming to "should be removed" or something more explicit ?

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 18]
```diff
@@ -26,15 +29,77 @@ export interface MovieData {
...
+    if(!licensee.orgId) {
+      delete licensee.orgId
+    }
+    if(typeof licensee.showName !== 'boolean') {
+      return false;
+    }
+  }
+
+  for(const licensor of licensors) {
+    if(!licensor.orgId) {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Same here

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts [hunk 19]
```diff
@@ -26,15 +29,77 @@ export interface MovieData {
...
+  // First, contract must have at least a licensee and a licensor
+
+  if (contract.parties.length < 2) { return false; }
+  const licensees = contract.parties.filter(p => p.role === getCodeIfExists('LEGAL_ROLES', 'licensee'))
+  const licensors = contract.parties.filter(p => p.role === getCodeIfExists('LEGAL_ROLES', 'licensor'))
+
+  if (!licensees.length || !licensors.length) { return false; }
+
+  for(const licensee of licensees) {
+    if(!licensee.orgId) {
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: I'm not sure to understand this line : if there is no "orgId", then delete "orgId".

File: apps/catalog/marketplace/marketplace/src/app/movie/home/home.component.html [hunk 20]
```diff
@@ -47,7 +47,7 @@ <h2>{{ movie.main.title.international }}</h2>
                       {{ country | translateSlug: 'TERRITORIES'}}{{isLast ? '' : ', '}}{{i == 1 && movie.main.originCountries.length > 2 ? ',  ' : ''}}
                     </span>
                     <span *ngIf="movie.main.originCountries.length > 2">... </span>
-                    <span *ngIf="movie.main.length">| {{ movie.main.length }} min </span>
+                    <span *ngIf="movie.main.totalRunTime">| {{ movie.main.totalRunTime }} min </span>
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/movie/home/home.component.html]: I'm not sure, but many more readable like that ? What do you think ?
```html
<span *ngIf="movie.main.totalRunTime as time">| {{ time }} min </span>
```

Comment (GrandSchtroumpf) [reply apps/catalog/marketplace/marketplace/src/app/movie/home/home.component.html]: You can apply in the other files if you think it's better.

File: apps/catalog/marketplace/marketplace/src/app/movie/view/view.component.html [hunk 21]
```diff
@@ -15,10 +15,10 @@ <h1>{{movie.main.title.international}}</h1>
         <div  class="movie-info" fxLayout="row" fxLayoutAlign="space-between center">
           <div>
             <span *ngIf="movie.main.productionYear">{{ movie.main.productionYear }}</span>
-            <span *ngIf="movie.main.status || movie.main.length"> | </span>
+            <span *ngIf="movie.main.status || movie.main.totalRunTime"> | </span>
```
Comment (GrandSchtroumpf) [inline apps/catalog/marketplace/marketplace/src/app/movie/view/view.component.html]: It's kind of difficult to understand what happen here. Would it make sense to create a pipe for that ?

File: libs/movie/src/lib/movie/+state/movie.model.ts [hunk 22]
```diff
@@ -227,17 +230,13 @@ export function createDistributionDeal(params: Partial<DistributionDeal> = {}):
...
-    price: createPrice(params.price),
     titleInternalAlias: '',
     formatProfile: FormatProfile.unknown,
     download: false,
     holdbacks: [],
-    fees: [],
-    ...params,
-    licensee: createStakeholder(params.licensee),
-    licensor: createStakeholder(params.licensor),
+    catchUp: createTerms(params.catchUp),
```
Comment (GrandSchtroumpf) [inline libs/movie/src/lib/movie/+state/movie.model.ts]: In the other factories you put the subcreate after the `...params`. Any reason to do it differently here ?

File: libs/utils/src/lib/common-interfaces/identity.ts [hunk 23]
```diff
@@ -81,6 +81,15 @@ export function createStakeholder(params: Partial<Stakeholder> = {}): Stakeholde
   }
 }
 
+export function createParty(params: Partial<Party> = {}): Party {
+  return {
+    ...createStakeholder(params),
```
Comment (GrandSchtroumpf) [inline libs/utils/src/lib/common-interfaces/identity.ts]: So a "Party" is a superset of the stakeholder ?

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts (hunk 5 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.firestore.ts]: roger that

File: apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts (hunk 19 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/distribution-deal/+state/cart.model.ts]: Lol. Thanks ..

File: apps/catalog/marketplace/marketplace/src/app/movie/home/home.component.html (hunk 20 above)
Comment (phpgeekfr) [reply apps/catalog/marketplace/marketplace/src/app/movie/home/home.component.html]: I agree

File: libs/movie/src/lib/movie/+state/movie.model.ts (hunk 22 above)
Comment (phpgeekfr) [reply libs/movie/src/lib/movie/+state/movie.model.ts]: in certain circumstances `...params` must be placed before subcreates because subcreates functions output can be erased right after creation by `...params`.
For example:
No need to put `...params` before:
```
export function createTerms(params: Partial<Terms> = {}): Terms {
  return {
    ...params
  };
}
```
but for this yes
```
export function createImgRef(ref: Partial<ImgRef> | string = {}): ImgRef {
  const _ref = typeof ref === 'string' ? { url: ref } : ref;
  return {
    url: '',
    ref: '',
    originalRef: '',
    ..._ref
  };
}
```

File: libs/utils/src/lib/common-interfaces/identity.ts (hunk 23 above)
Comment (phpgeekfr) [reply libs/utils/src/lib/common-interfaces/identity.ts]: Party is a Person with a legalRole. Identity => Person => Party

image
